### PR TITLE
Api Tag Endpoint Feature Complete With Tests

### PR DESCRIPTION
**Pull request by HankG**
WIP PR that needs to be collapsed to one commit after review

### CODE REVIEW

File: app/controllers/api/v1/tag_followings_controller.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        render json: tag_followings_service.index.map(&:name)
+      end
+
+      def create
+        tag_followings_service.create(params.require(:name))
+        head :no_content
+      end
+
+      def destroy
+        tag_followings_service.destroy_by_name(params.require(:id))
```
Comment (frankrousseau) [inline app/controllers/api/v1/tag_followings_controller.rb:28]: The name `destroy_by_name` is a bit confusing when you use a parameter named `id`

File: app/controllers/tag_followings_controller.rb [hunk 2]
```diff
@@ -14,28 +14,18 @@ class TagFollowingsController < ApplicationController
...
-      if @tag_following.save
-        render :json => @tag.to_json, :status => 201
-      else
-        head :forbidden
-      end
-    end
+    tag = tag_followings_service.create(params["name"])
+    render json: tag.to_json, status: :created
+  rescue StandardError
+    head :forbidden
```
Comment (frankrousseau) [inline app/controllers/tag_followings_controller.rb:20]: I think we should not change anything which is not related to the API.

File: app/services/tag_following_service.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    tag = ActsAsTaggableOn::Tag.find_or_create_by(name: name_normalized)
+    tag_following = @user.tag_followings.new(tag_id: tag.id)
+
+    raise "Can't process tag entity" unless tag_following.save
+    tag
+  end
+
+  def destroy(id)
+    tag_following = @user.tag_followings.find_by(tag_id: id)
+    tag_following&.destroy
```
Comment (frankrousseau) [inline app/services/tag_following_service.rb:21]: Are you sure about the `&` syntax?

Comment (frankrousseau) [reply app/services/tag_following_service.rb:21]: I don't know Ruby well, that's why I'm asking.

File: spec/controllers/tag_followings_controller_spec.rb [hunk 4]
```diff
@@ -29,4 +29,67 @@
...
+      expect(response.status).to be(201)
+      tag_data = JSON.parse(response.body)
+      expect(tag_data["name"]).to eq(name)
+      expect(tag_data.has_key?("id")).to be_truthy
+      expect(tag_data["taggings_count"]).to eq(0)
+    end
+
+    it "Fails with missing name field" do
+      post :create, params: {}
+      expect(response.status).to eq(403)
```
Comment (frankrousseau) [inline spec/controllers/tag_followings_controller_spec.rb:50]: Missing params should return 400 error.

File: spec/controllers/tag_followings_controller_spec.rb [hunk 5]
```diff
@@ -29,4 +29,67 @@
...
+
+    it "Deletes tag with valid id" do
+      delete :destroy, params: {id: @tag_id_to_delete}, format: :json
+      expect(response.status).to eq(204)
+      expect(alice.followed_tags.find_by(name: @tag_name)).to be_nil
+    end
+
+    it "Fails with missing name field" do
+      delete :create, params: {}, format: :json
+      expect(response.status).to eq(403)
```
Comment (frankrousseau) [inline spec/controllers/tag_followings_controller_spec.rb:70]: Same here. 403 usually is related to authorization errors.

File: spec/controllers/tag_followings_controller_spec.rb [hunk 6]
```diff
@@ -29,4 +29,67 @@
...
+    end
+
+    it "Fails with bad Tag ID" do
+      delete :create, params: {id: -1}, format: :json
+      expect(response.status).to eq(403)
```
Comment (frankrousseau) [inline spec/controllers/tag_followings_controller_spec.rb:75]: Same here. 403 usually is related to authorization errors.

File: app/controllers/api/v1/tag_followings_controller.rb (hunk 1 above)
Comment (HankG) [reply app/controllers/api/v1/tag_followings_controller.rb:28]: We can probably call the field "name" or something since that's what is being sent in.  I just went with the default off the path parser.  Let me look into it.

File: app/controllers/tag_followings_controller.rb (hunk 2 above)
Comment (HankG) [reply app/controllers/tag_followings_controller.rb:20]: I have to though to get the tag processing into a service component and then have both use the service component as we discussed.

File: app/services/tag_following_service.rb (hunk 3 above)
Comment (HankG) [reply app/services/tag_following_service.rb:21]: It's the safe navigation operator so should be replacing logic like "tag_following && tag_following.destroy"  While all the unit tests pass thing may need to look into is if it should be returning false instead of nil or something like that, but with all the unit tests passing even for error conditions I thought it was done correctly.

File: spec/controllers/tag_followings_controller_spec.rb (hunk 4 above)
Comment (HankG) [reply spec/controllers/tag_followings_controller_spec.rb:50]: On the 403 vs 400 error, there was nothing in the spec about error codes for this endpoint and others so I was trying to make it consistent with the endpoints that do have them.  They were using either 403 or 422 errors.  Probably 422 would be better if we want to reuse an existing code.  We could submit with a 400 error and see if they balk when I make the documentation change.

File: app/controllers/tag_followings_controller.rb (hunk 2 above)
Comment (frankrousseau) [reply app/controllers/tag_followings_controller.rb:20]: Ok. I hope it won't bother maintainers.

File: app/services/tag_following_service.rb (hunk 3 above)
Comment (frankrousseau) [reply app/services/tag_following_service.rb:21]: That's great. Thank you for the explaination. I didn't know that operator.

File: spec/controllers/tag_followings_controller_spec.rb (hunk 4 above)
Comment (frankrousseau) [reply spec/controllers/tag_followings_controller_spec.rb:50]: Forget it. It is not part of the API entpoints. We should not change that or it will lead too to many changes for the final merge. Sorry.

File: app/controllers/api/v1/tag_followings_controller.rb (hunk 1 above)
Comment (frankrousseau) [reply app/controllers/api/v1/tag_followings_controller.rb:28]: Tell me when you take your decision, I will merge your PR after that.

Comment (HankG) [reply app/controllers/api/v1/tag_followings_controller.rb:28]: I'm not finding a way to override the name in Rails.  However the value that is in that field is the name not a Tag ID, so the method name is correct.  When I look at where I thought it was done it was an artifact of Rails providing different names a long a path so that a path of `posts/1234/comments/1` gets two parameters of posts_id and id.  If you want to squash and merge as is that works for me.

Comment (frankrousseau) [reply app/controllers/api/v1/tag_followings_controller.rb:28]: Ok I got it.

**Comment by frankrousseau** [comment]
Let's go for squashing!

**Comment by HankG** [comment]
Do you want to squash this and the other before merging or would you like me to?

**Comment by frankrousseau** [comment]
It's better if you do it.
